design/proposal-concurrent-collectors.md

```diff
@@ -0,0 +1,207 @@
+
+Ensure that when needed, a collector can ensure that another collector has run prior to starting without needing to rely on the ordering in a spec.
+
+## Non Goals
+
+A large portion of this change involves the logs collector, while this is not the primary goal it does allow collectors to trigger pod log collection from one place.
+ 
+## Background
+
+Currently all collectors run sequentially, in the order in which they are specified in the spec.  With the addition of the ability to supply multiple specs, the order is no longer guaranteed.
+
+The collector code includes a check to ensure that `clusterResources` and `clusterInfo` are added to every collection regardless of the spec.  These are added at the end of the run, and so collect data for pods created during the Troubleshoot collection, which causes poor output results (see [#767](https://github.com/replicatedhq/troubleshoot/issues/767)).
+
+## High-Level Design
+
+Introduce a new parameter for `supportBundle` spec collectors: `dependsOn`.  This is to be used to ensure that a specific collector will only run once the one specified is completed.  e.g.:
+
+```yaml
+    - exec:
+        name: generate-output
+        collectorName: generate-output
+        namespace: '{{repl Namespace}}'
+        selector:
+          - app=some-selector
+        command: ["sh"]
+        args: ["-c", "cd /some/data/dir && mv `ls -t | head -n 1` latest"]
+        timeout: 5s
+    - copy:
+        name: collect-output
+        dependsOn: generate-output
+        collectorName: collect-output
+        selector:
+          - app=some-selector
+        namespace: '{{repl Namespace}}'
+        containerPath: /some/data/dir/latest
+```
+
+This concept requires
+
+* error handling (what if that collector isn't in the spec, or fails)
+* logic to order all the collectors that depend on others as there might be several in the line
+
+Introduce a new function(s) which does the work of pod log collection, storing those logs where `sbctl` can find them in `/cluster-resources/pods/logs/[namespace]/[pod]/[container].log`.  This should start in a goroutine, prior to any collectors being run.
+
+Switch all collectors to use goroutines to collect (concurrently).  This should be written such that collectors are split into separate lists, in order that any collectors with `dependsOn` are in a phase later than the collector it depends on.  Run each phase in turn, at this stage it may be acceptable to wait for the entire list of collectors in that phase to complete prior to starting the next phase.
+
+Add a channel to each collector (including `logs`), of a type where collectors can, towards the end of their run, send a slice to the channel specifying the pods that particular collector wants to collect logs from.  For the `logs` and `clusterResources` collectors, this replaces the actual pod log collection routine.
```

Comment:
I think this might be overcomplicating things, it might be simpler to initialise the collector goroutine with everything it needs up front and leave the channel for communicating back to the parent process.

Reply:
+1

I think having a single place where any collector can say "collect logs from this pod" is probably much simpler compared to sharing a channel all to all collectors.